### PR TITLE
Implementation of insert and replaceifsafe

### PR DESCRIPTION
**Pull request by yonathanF**
Opening the PR so we can talk about the code while we test it and fix things. 

I still need to test it and I will leave comments for the places I think will need attention

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,3 +12,7 @@ test_run_dir
 *~
 \#*\#
 .\#*
+.bloop/
+.metals/
+.scalafmt.conf
```
Comment (yonathanF) [inline .gitignore:17]: We might want to consider actually using a formatter for consistency

File: core/src/main/scala/chisel3/SimpleChiselTransformer.scala [hunk 2]
```diff
@@ -6,91 +6,178 @@ import chisel3.internal.firrtl._
...
+  }
+
+  private def dependsOnArg(cmd: Command, eql: Arg => Boolean): Boolean =
+    cmd match {
+      case Connect(_, _, exp)     => eql(exp)
+      case WhenBegin(_, pred)     => eql(pred)
+      case ConnectInit(_, _, exp) => eql(exp)
+      case Stop(_, clock, _)      => eql(clock)
+      case Printf(_, clock, _)    => eql(clock)
+      case _                      => false
```
Comment (yonathanF) [inline core/src/main/scala/chisel3/SimpleChiselTransformer.scala]: Should we be checking if definition commands depend on that id?

File: core/src/main/scala/chisel3/SimpleChiselTransformer.scala [hunk 3]
```diff
@@ -6,91 +6,178 @@ import chisel3.internal.firrtl._
...
     }
 
-    def removeIfSafe(id: Arg, ctx: ArrayBuffer[Command]): Option[ArrayBuffer[Command]] ={
-        //TODO
-        return None   
+  private def getDefinitionUid(command: Command): Option[BigInt] =
+    command match {
+      // Right(args.find(a => eql(a)).map(a => a.uniqueId)) what's the id here?
+      case DefPrim(_, _, _, args @ _*)  => Some(args(0).uniqueId)
```
Comment (yonathanF) [inline core/src/main/scala/chisel3/SimpleChiselTransformer.scala]: for DefPrim types, what is considered the ID? I just defaulted it to the first for now but clearly this isn't right

File: core/src/main/scala/chisel3/SimpleChiselTransformer.scala [hunk 4]
```diff
@@ -6,91 +6,178 @@ import chisel3.internal.firrtl._
...
-        //TODO
-        return None   
+  def removeIfSafe(
+      id: Arg,
+      ctx: ArrayBuffer[Command]
+  ): Option[ArrayBuffer[Command]] = {
+
+    def argEquality(a: Arg) = a.uniqueId.equals(id.uniqueId)
+
+    def matchesUID(c: Command) = getDefinitionUid(c) match {
```
Comment (yonathanF) [inline core/src/main/scala/chisel3/SimpleChiselTransformer.scala]: This doesn't feel idiomatic to me. basically want the truth value of the Option when the function is applied to it. Might be better done with a map like this: 

`      getDefinitionUid(c).map(a => a.equals(id.uniqueId)).getOrElse(false)
`

File: core/src/main/scala/chisel3/SimpleChiselTransformer.scala (hunk 2 above)
Comment (yonathanF) [reply core/src/main/scala/chisel3/SimpleChiselTransformer.scala]: Def reg init needs to object needs to be checked

File: core/src/main/scala/chisel3/SimpleChiselTransformer.scala (hunk 3 above)
Comment (yonathanF) [reply core/src/main/scala/chisel3/SimpleChiselTransformer.scala]: it has an id arg (second param) take that

File: core/src/main/scala/chisel3/SimpleChiselTransformer.scala [hunk 5]
```diff
@@ -6,91 +6,178 @@ import chisel3.internal.firrtl._
...
+      cmd: Command,
+      ctx: ArrayBuffer[Command]
+  ): Option[ArrayBuffer[Command]] = {
+    // is the command a def, if so all its uses need to come after
+    // if it uses other nodes, these need to be before it
+
+    val ourFirstDependent = getDefinitionUid(cmd) match {
+      case Some(uid) =>
+        ctx
+          .filter(c => dependsOnArg(cmd, a => a.equals(uid)))
```
Comment (yonathanF) [inline core/src/main/scala/chisel3/SimpleChiselTransformer.scala]: cmd shoudl be c

File: core/src/main/scala/chisel3/SimpleChiselTransformer.scala [hunk 6]
```diff
@@ -6,91 +6,178 @@ import chisel3.internal.firrtl._
...
+          .map(c => ctx.indexOf(c))
+          .lift(0)
+          .map(_ + 1)
```
Comment (yonathanF) [inline core/src/main/scala/chisel3/SimpleChiselTransformer.scala]: should be -

File: core/src/main/scala/chisel3/SimpleChiselTransformer.scala [hunk 7]
```diff
@@ -6,91 +6,178 @@ import chisel3.internal.firrtl._
...
+      val new_ctx = new ArrayBuffer[Command]
+      return Some(new_ctx)
+    } else {
+      return Some(ctx)
+    }
+  }
+
+  private def dependsOnArg(cmd: Command, eql: Arg => Boolean): Boolean =
+    cmd match {
+      case Connect(_, _, exp)     => eql(exp)
```
Comment (yonathanF) [inline core/src/main/scala/chisel3/SimpleChiselTransformer.scala]: All nodes need to be considered for dependency. 
def prim needs to be considered (all args)

**Review by shibo-chen — APPROVED** [review]

**Pull request merged**